packages/grpc-native-core/index.d.ts

```diff
@@ -132,6 +132,8 @@ declare module "grpc" {
    */
   export function setLogVerbosity(verbosity: logVerbosity): void;
 
+  export type Service<T> = ProtobufService;
```

Comment:
What is the purpose of a generic type that doesn't use the type parameter in its definition?

Reply:
This gives the Ts compiler the ability to infer the type in `Server.addService()`:

```ts
addService<T>(service: Service<T>, implementation: T): void;
```
Since you give a `Service<T>` object as the first argument, it follows that the implementation in the second argument must satisfy `T`. Thus you usually do not need to add the parameter at all. For example:

```ts
// definitions
type RouteGuide = typeof routeguide.Client;

namespace routeguide {
	interface Service {
		getFeature(call: ServerUnaryCall<Point>, callback: sendUnaryData<Feature>): void;
		listFeatures(call: ServerWriteableStream<Rectangle>): void;
		recordRoute(call: ServerReadableStream<Point>, callback: sendUnaryData<RouteSummary>): void;
		routeChat(call: ServerDuplexStream<RouteNote, RouteNote>): void;
	}

	declare class Client extends GrpcClient {
		static service: GrpcService<Service>;

		getFeature(arg: Point, callback: sendUnaryData<Feature>): void;
		listFeatures(arg: Rectangle): ClientReadableStream<Feature>;
		recordRoute(callback: sendUnaryData<RouteSummary>): ClientWritableStream<Point>;
		routeChat(): ClientDuplexStream<RouteNote, RouteNote>;
	}
}

// usage

server.addService(RouteGuide.service, {
	getFeature(call: ServerUnaryCall<Point>, callback: sendUnaryData<Feature>): void {
		// omitted
	},
	listFeatures(call: ServerWriteableStream<Rectangle>): void {
		// omitted
	},
	recordRoute(call: ServerReadableStream<Point>, callback: sendUnaryData<RouteSummary>): void {
		// omitted
	},
	routeChat(call: ServerDuplexStream<RouteNote, RouteNote>): void {
		// omitted
	}
});
```

Reply:
It looks like what you're saying here is that the user has to both define `T` and explicitly state that another object has the type `Service<T>`. I don't see how that improves type safety or provides any benefit whatsoever.

Reply:
If you have a `T` defined (preferably autogenerated from the proto files) like in the example above, then it is an improvement, since the compiler will warn you about any missing/wrong function declarations. Also you do not need to create a `Service<T>` type, you only need 
to define the server side API (`T`) like in `routeguide.Service` and `GrpcService<Service>` will give you the correct type.

We could give `T` a default value, so users do not have to provide a type definition like so:
```
type ServiceImplementation = { [name: string]: handleCall };

export class Server {
    // …
    addService<T = ServiceImplementation>(service: Service<T>, implementation: T): void;
}
```
But having a generic implementation definition actually means that you cannot benefit from Typescripts type checking.